### PR TITLE
SSCS-10978: initial implementation of send letter functionality

### PR DESCRIPTION
**Pull request by AndreiTrushkov**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/SSCS-10978

### Change description ###
initial implementation of send letter functionality for generic letter

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by jarekPierchala — APPROVED** [review]

**Review by hajam09 — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+        docmosisTemplate = getDocmosisTemplate(caseData);
+        docmosisCoverSheetTemplate = getDocmosisCoverSheet(caseData);
+
+        process(id, caseData);
+    }
+
+    private void sendLetter(long caseId, SscsCaseData caseData, List<Pdf> pdfs) {
+        log.info("Sending letter");
+        Optional<UUID> id = bulkPrintService.sendToBulkPrint(pdfs, caseData);
+        //Optional.of(UUID.randomUUID());
```
Comment (jarekPierchala) [inline src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java:110]: remove that comment

File: src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+    @Override
+    public void handle(CallbackType callbackType, Callback<SscsCaseData> callback) {
+        if (!canHandle(callbackType, callback)) {
+            log.info("Cannot handle this event for case id: {}", callback.getCaseDetails().getId());
+            throw new IllegalStateException("Cannot handle callback");
+        }
+
+        SscsCaseData caseData = callback.getCaseDetails().getCaseData();
+        long id = callback.getCaseDetails().getId();
```
Comment (jarekPierchala) [inline src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java:99]: caseDetailsId as a name please

File: src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+        if (YesNo.isYes(caseData.getSendToJointParty())) {
+            sendToJointParty(caseId, caseData, documents);
+        }
+
+        if (YesNo.isYes(caseData.getSendToOtherParties())) {
+            sendToOtherParties(caseId, caseData, documents);
+        }
+
+        // blank page??
```
Comment (jarekPierchala) [inline src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java:162]: if its something to be done leave TODO comment rather then what you have got now

File: src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+            }
+        }
+    }
+
+    private FurtherEvidenceLetterType getLetterType(OtherParty otherParty, String entityId) {
+        if (otherParty.hasRepresentative() && entityId.equals(otherParty.getRep().getId())) {
+            return FurtherEvidenceLetterType.OTHER_PARTY_REP_LETTER;
+        }
+
+        return FurtherEvidenceLetterType.OTHER_PARTY_LETTER;
```
Comment (jarekPierchala) [inline src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java:190]:
```suggestion
       boolean hasRepresentative = otherParty.hasRepresentative() && entityId.equals(otherParty.getRep().getId());
       return hasRepresentative ? FurtherEvidenceLetterType.OTHER_PARTY_REP_LETTER :FurtherEvidenceLetterType.OTHER_PARTY_LETTER;
```

File: src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        return letter;
+    }
+
+    private byte[] buildBundledLetter(byte[] coverSheet, byte[] letter) {
+        if (coverSheet != null) {
+            PDDocument bundledLetter = null;
+
+            try {
+                bundledLetter = PDDocument.load(letter);
+
```
Comment (jarekPierchala) [inline src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java:273]: no need for that empty line

File: src/main/java/uk/gov/hmcts/reform/sscs/service/CoverLetterService.java [hunk 6]
```diff
@@ -101,4 +110,63 @@ public byte[] generateCoverLetterRetry(FurtherEvidenceLetterType letterType, Str
...
+    public List<Pdf> getSelectedDocuments(SscsCaseData sscsCaseData) {
+        List<Pdf> documents = new ArrayList<>();
+
+        for (CcdValue<DocumentSelectionDetails> d : sscsCaseData.getDocumentSelection()) {
+            var documentLink = findDocumentByFileName(d.getValue().getDocumentsList().getValue().getCode(), sscsCaseData);
+            byte[] document = null;
+
+            if (documentLink != null) {
+                document = pdfStoreService.download(documentLink.getDocumentUrl());
+
```
Comment (jarekPierchala) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/CoverLetterService.java:134]: no need for empty line

File: src/main/java/uk/gov/hmcts/reform/sscs/service/placeholders/GenericLetterPlaceholderService.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        String benefitAcronym = benefit.isHasAcronym() ? benefit.getShortName() : benefit.getDescription();
+        placeholders.put(BENEFIT_NAME_ACRONYM_LITERAL, benefitAcronym);
+
+        placeholders.put(SSCS_URL_LITERAL, SSCS_URL);
+        placeholders.put(GENERATED_DATE_LITERAL, LocalDateTime.now().toLocalDate().toString());
+        placeholders.put(IS_REPRESENTATIVE, "No");
+
+        if (isRepresentativeLetter(letterType) || isOtherPartyLetter(letterType)) {
+            String representativeName = PlaceholderUtility.getName(caseData, letterType, otherPartyId);
+
```
Comment (jarekPierchala) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/placeholders/GenericLetterPlaceholderService.java:81]: no need for empty line

File: src/main/java/uk/gov/hmcts/reform/sscs/service/placeholders/GenericLetterPlaceholderService.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        }
+        if (lines.length >= 3) {
+            addressPlaceHolders.put(LETTER_ADDRESS_LINE_3, truncateAddressLine(defaultToEmptyStringIfNull(lines[2])));
+        }
+        if (lines.length >= 4) {
+            addressPlaceHolders.put(LETTER_ADDRESS_LINE_4, truncateAddressLine(defaultToEmptyStringIfNull(lines[3])));
+        }
+        if (lines.length >= 5) {
+            addressPlaceHolders.put(LETTER_ADDRESS_POSTCODE, truncateAddressLine(defaultToEmptyStringIfNull(lines[4])));
+        }
```
Comment (jarekPierchala) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/placeholders/GenericLetterPlaceholderService.java:144]:
```suggestion
         var addressPlaceHolders = new HashMap<String, Object>();
        String[] lines = lines(address);
        String[] addressConstants = {LETTER_ADDRESS_LINE_1, LETTER_ADDRESS_LINE_2, LETTER_ADDRESS_LINE_3, LETTER_ADDRESS_LINE_4, LETTER_ADDRESS_POSTCODE};

        for (int i = 0; i <= lines.length; i++){
            addressPlaceHolders.put(addressConstants[i], truncateAddressLine(defaultToEmptyStringIfNull(lines[i])));
        }
``` 

Try this out Andrei

File: src/main/java/uk/gov/hmcts/reform/sscs/service/placeholders/PlaceholderUtility.java [hunk 9]
```diff
@@ -13,4 +33,142 @@ static String defaultToEmptyStringIfNull(String value) {
...
+            .filter(jpn -> isValidName(Name.builder().firstName(jpn.getFirstName()).lastName(jpn.getLastName()).build()))
+            .map(Name::getFullNameNoTitle)
+            .orElse(SIR_MADAM);
+    }
+
+    private static String getOtherPartyName(SscsCaseData caseData, String otherPartyId) {
+        if (otherPartyId != null) {
+            for (CcdValue<OtherParty> otherParty : caseData.getOtherParties()) {
+                if (otherPartyId.contains(otherParty.getValue().getId())
+                    && isValidName(otherParty.getValue().getName())) {
```
Comment (jarekPierchala) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/placeholders/PlaceholderUtility.java:155]: otherParty.getValue() to seperate var

**Review by jarekPierchala — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java (hunk 1 above)
Comment (AndreiTrushkov) [reply src/main/java/uk/gov/hmcts/reform/sscs/callback/handlers/IssueGenericLetterHandler.java:110]: done

**Review by jarekPierchala — APPROVED** [review]

**Review by jarekPierchala — APPROVED** [review]

**Pull request merged**